### PR TITLE
Live streaming integration

### PR DESCRIPTION
**Pull request by justinswart**
This PR adds the following:

### Within the app:
- Live streaming!
- `LiveStreamContainerViewController` contains the live video rendering view controller as well as its accompanying view controller which is currently only displaying details about the live stream but which will in future provide live chat.
- `LiveStreamCountdownViewController` is displayed just before a stream goes live with some basic details about the upcoming stream, creator and a timer that ticks over. When it reaches zero then it pushes the `LiveStreamContainerViewController` and replaces itself in the navigation stack.
- These two view controllers have the view models `LiveStreamContainerViewModel` and `LiveStreamCountdownViewModel` respectively. They also both share the `LiveStreamEventDetailsViewModel` as most of this information is displayed on both views.
- `LiveStream.storyboard` the story board for the aforementioned view controllers.
- `ClearNavigationBar` a super simple `UINavigationBar` subclass that provides the necessary styling for the nav bar on the live stream views.
- `LiveStream` framework, the live streaming framework which contains the Firebase and OpenTok dependencies and live video view controllers - note that this was previously a submodule but has since been moved into the app as a framework.

It also makes the following changes:

- Refactors `ProjectPamphletSubpageCell` to use a view model: `ProjectPamphletSubpageCellViewModel`. This felt like a necessary change after introducing a third subpage type for live streams.  Comments and Updates were very similar so I think one could still get away without the testable inputs/outputs of a view model but live streams can be in several different states so I wanted it to be testable.
- Adds one or two additional button styles.
- Adds push notification paths for live stream deep-linking.

### LiveStream framework:
- Refactored live streaming view controllers (these previously had more UI components but we've since decided that those belong in the main app to make use of all of the UIKit extensions and style helpers). This also means that there isn't that much delegate crosstalk needed between the framework and the app, it's really just the playback state that we share and the event id that is passed in from the app.
- `LiveVideoViewController` a basic video rendering view controller that can display OpenTok and HLS video streams.
- `LiveStreamViewController` the view controller that manages a nested aforementioned `LiveVideoViewController` and listens to Firebase events for things like green room or HLS url, etc. Green room is a boolean that is activated/deactivated when the creator decides to go live or not. We use this listener to destroy/recreate the video view.
- Tests for these controller's view models.
- `LiveStreamEvent` model which is the event object that we return from our live.kickstarter.com events feed.
- Lenses for `LiveStreamEvent`.

### Screenshots:
`LiveStreamCountdownViewController`:

teststandardview_lang_en_device_phone4_7inch_orientation_portrait 2x

`LiveStreamContainerViewController`:

testlive_lang_en_device_phone4_7inch_orientation_portrait 2x

`ProjectPamphletContentViewController`:

testnonbacker_liveproject_withlivestreams_lang_en_device_phone4_7inch 2x

Browse screenshots of the countdown [here](https://github.com/kickstarter/ios-oss/tree/live_streaming_integration/Screenshots/_64/Kickstarter_Framework_iOSTests.LiveStreamCountdownViewControllerTests) and live stream container [here](https://github.com/kickstarter/ios-oss/tree/live_streaming_integration/Screenshots/_64/Kickstarter_Framework_iOSTests.LiveStreamContainerViewControllerTests). Project Pamphlet [here](https://github.com/kickstarter/ios-oss/tree/live_streaming_integration/Screenshots/_64/Kickstarter_Framework_iOSTests.ProjectPamphletContentViewControllerTests).

### CODE REVIEW

File: Kickstarter-iOS/AppDelegate.swift [hunk 1]
```diff
@@ -40,7 +41,7 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
     AppEnvironment.replaceCurrentEnvironment(facebookAppDelegate: FBSDKApplicationDelegate.sharedInstance())
 
     #if DEBUG
-      if Secrets.isOSS {
+      if KsApi.Secrets.isOSS {
```
Comment (mbrandonw) [inline Kickstarter-iOS/AppDelegate.swift]: so both `LiveStream` and `KsApi` frameworks have access to the secrets, and so we need to disambiguate that. luckily both frameworks have exactly the same secrets...

File: Kickstarter-iOS/Views/Controllers/LiveStreamCountdownViewController.swift [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+  @IBOutlet private weak var liveStreamParagraphLabel: UILabel!
+  @IBOutlet private weak var minutesSubtitleLabel: UILabel!
+  @IBOutlet private weak var minutesTitleLabel: UILabel!
+  @IBOutlet private weak var projectImageView: UIImageView!
+  @IBOutlet private weak var secondsSubtitleLabel: UILabel!
+  @IBOutlet private weak var secondsTitleLabel: UILabel!
+  @IBOutlet private weak var subscribeActivityIndicatorView: UIActivityIndicatorView!
+  @IBOutlet private weak var subscribeButton: UIButton!
+
+  private let eventDetailsViewModel: LiveStreamEventDetailsViewModelType = LiveStreamEventDetailsViewModel()
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Controllers/LiveStreamCountdownViewController.swift]: Both the `LiveStreamCountDownViewController` and `LiveStreamContainerViewController` share this `LiveStreamEventDetailsViewModel`. It's perhaps not the most ideal situation but there is a fair bit of code duplication between the two view controllers so I chose to do it this way.

One thing to bear in mind is that the `LiveStreamEventDetailsViewModel` fetches the `LiveStreamEvent` model and updates the other view models with this.

File: Kickstarter-iOS/Library/UIImageView+URL.swift [hunk 3]
```diff
@@ -15,3 +18,30 @@ extension UIImageView {
                      completion: nil)
   }
 }
+
+private enum Associations {
+  fileprivate static var imageUrl = 0
+}
+
+extension Rac where Object: UIImageView {
```
Comment (mbrandonw) [inline Kickstarter-iOS/Library/UIImageView+URL.swift]: I don't know why we never did this before, but this is a reactive extension so that we can just do 

```swift
self.imageView.rac.imageUrl = self.viewModel.outputs.imageUrl
```

without all the cancel/set to nil/load image hoopla we've done all over the place.

File: Kickstarter-iOS/ViewModels/AppDelegateViewModel.swift [hunk 4]
```diff
@@ -831,3 +851,23 @@ extension ShortcutItem {
     }
   }
 }
+
+private func liveStreamData(fromNavigation nav: Navigation)
+  -> SignalProducer<(Project, Project.LiveStream, LiveStreamEvent, RefTag?), NoError> {
+
+    guard case let .project(projectParam, .liveStream(eventId), refTag) = nav else { return .empty }
+
+    return SignalProducer.zip(
```
Comment (mbrandonw) [inline Kickstarter-iOS/ViewModels/AppDelegateViewModel.swift]: this is a fun use of `zip`! we need to call out to the main kickstarter api and the live stream api to fetch some data, so might as well do it concurrently!

File: Kickstarter-iOS/Views/Cells/ProjectPamphletMainCell.swift [hunk 5]
```diff
@@ -113,9 +113,7 @@ internal final class ProjectPamphletMainCell: UITableViewCell, ValueCell {
       |> UILabel.lens.numberOfLines .~ 2
 
     _ = self.creatorButton
-      |> UIButton.lens.accessibilityHint %~ { _ in
-        localizedString(key: "Opens_creator_profile", defaultValue: "Opens creator profile.")
-      }
+      |> UIButton.lens.accessibilityHint %~ { _ in Strings.Opens_creator_profile() }
```
Comment (mbrandonw) [inline Kickstarter-iOS/Views/Cells/ProjectPamphletMainCell.swift]: cleaned up some strings while i was in here

File: Kickstarter-iOS/Views/Cells/ProjectPamphletSubpageCell.swift [hunk 6]
```diff
@@ -3,61 +3,21 @@ import Library
 import Prelude
 import UIKit
 
-internal enum ProjectPamphletSubpage {
```
Comment (mbrandonw) [inline Kickstarter-iOS/Views/Cells/ProjectPamphletSubpageCell.swift]: all of this moved to the vm where it got built out a bit more

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift [hunk 7]
```diff
@@ -12,7 +12,7 @@ internal final class CommentsViewController: UITableViewController {
 
   // This button needs to store a strong reference so as to not get wiped when setting hidden state.
   @IBOutlet fileprivate var commentBarButton: UIBarButtonItem!
-  fileprivate weak var loginToutViewController: UIViewController? = nil
+  fileprivate weak var loginToutViewController: UIViewController?
```
Comment (mbrandonw) [inline Kickstarter-iOS/Views/Controllers/CommentsViewController.swift]: new swiftlint rule caught this ¯\\_(ツ)_/¯

File: Library/Koala/Koala.swift [hunk 8]
```diff
@@ -1790,6 +1809,9 @@ private func properties(shareContext: ShareContext,
   case let .project(project):
     result = result.withAllValuesFrom(properties(project: project, loggedInUser: loggedInUser))
     result["context"] = "project"
+  case let .liveStream(project, _):
+    result = result.withAllValuesFrom(properties(project: project, loggedInUser: loggedInUser))
+    result["context"] = "liveStream"
```
Comment (mbrandonw) [inline Library/Koala/Koala.swift]: @justinswart we typically use snake casing for property values, so should prob be `live_stream`

File: Library/LoginIntent.swift [hunk 9]
```diff
@@ -20,6 +21,8 @@ public enum LoginIntent: String {
       return "favorite_category"
     case .generic:
       return "generic"
+    case .liveStreamSubscribe:
+      return "liveStreamSubscribe"
```
Comment (mbrandonw) [inline Library/LoginIntent.swift]: same here with underscores

File: Library/Tests/TestHelpers/TestCase.swift [hunk 10]
```diff
@@ -15,8 +16,9 @@ internal class TestCase: FBSnapshotTestCase {
   internal let cookieStorage = MockCookieStorage()
   internal let dateType = MockDate.self
   internal let facebookAppDelegate = MockFacebookAppDelegate()
+  internal let liveStreamService = MockLiveStreamService(fetchEventResult: nil)
   internal let mainBundle = MockBundle()
-  internal let scheduler = TestScheduler()
+  internal let scheduler = TestScheduler(startDate: MockDate().date)
```
Comment (mbrandonw) [inline Library/Tests/TestHelpers/TestCase.swift]: we did some fun new time-based testing stuff in this PR, and this lil fella was the key to having our test scheduler and mock date stuff in sync. i'll call out some other cool stuff we did later...

File: Library/Tests/ViewModels/BackingCellViewModelTests.swift [hunk 11]
```diff
@@ -25,9 +25,9 @@ internal final class BackingCellViewModelTests: TestCase {
   }
 
   func testOutputs() {
-    let reward = .template |> Reward.lens.estimatedDeliveryOn .~ NSDate().timeIntervalSince1970
+    let reward = .template |> Reward.lens.estimatedDeliveryOn .~ Date().timeIntervalSince1970
```
Comment (mbrandonw) [inline Library/Tests/ViewModels/BackingCellViewModelTests.swift]: 🤔 weird... i guess these slipped by my swift 3 conversion...

File: Library/ViewModels/LiveStreamContainerViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        }
+      }
+
+    self.loaderText = Signal.merge(
+      liveStreamState.map {
+        if case .live(playbackState: .loading, _) = $0 { return Strings.The_live_stream_will_start_soon() }
+        if case .greenRoom = $0 { return Strings.The_live_stream_will_start_soon() }
+        if case .replay(playbackState: .loading, _) = $0 {
+          return Strings.The_replay_will_start_soon()
+        }
```
Comment (mbrandonw) [inline Library/ViewModels/LiveStreamContainerViewModel.swift]: @justinswart since we are `if case`ing this so much it might be better to just do a switch

Comment (mbrandonw) [inline Library/ViewModels/LiveStreamContainerViewModel.swift]: there are a few spots in this file we can do this

File: Library/Koala/Koala.swift (hunk 8 above)
Comment (justinswart) [inline Library/Koala/Koala.swift]: Cool, updated.

File: Library/LoginIntent.swift (hunk 9 above)
Comment (justinswart) [inline Library/LoginIntent.swift]: 👍

File: Library/ViewModels/LiveStreamContainerViewModel.swift (hunk 12 above)
Comment (justinswart) [inline Library/ViewModels/LiveStreamContainerViewModel.swift]: Good call, updated

File: Kickstarter-iOS/Library/UIImageView+URL.swift (hunk 3 above)
Comment (theginbin) [inline Kickstarter-iOS/Library/UIImageView+URL.swift]: don't we ideally want all images in the app to have a cross-dissolve though? (i assumed we'd change all `af_setImageWithURL` to `ksr_setImageWithURL` at some point.)

File: Kickstarter-iOS/DataSources/ProjectPamphletContentDataSource.swift [hunk 13]
```diff
@@ -52,6 +56,51 @@ internal final class ProjectPamphletContentDataSource: ValueCellDataSource {
...
+        : (lhs.startDate == rhs.startDate ? .eq : lhs.startDate < rhs.startDate ? .gt: .lt)
+    }
+
+    // Sort by:
+    //   * live streams first
+    //   * then future streams first and past streams last
+    //   * future streams sorted by start date asc, past streams sorted by start date desc
+    let comparator = currentlyLiveStreamsFirstComparator
+      <> futureLiveStreamsFirstComparator
+      <> startDateComparator
```
Comment (theginbin) [inline Kickstarter-iOS/DataSources/ProjectPamphletContentDataSource.swift]: this is wild. so clean. wow.

File: Kickstarter-iOS/Library/UIImageView+URL.swift (hunk 3 above)
Comment (theginbin) [inline Kickstarter-iOS/Library/UIImageView+URL.swift]: oh NM! i just saw the full code, excellent!

File: Kickstarter-iOS/DataSources/ProjectPamphletContentDataSource.swift (hunk 13 above)
Comment (mbrandonw) [inline Kickstarter-iOS/DataSources/ProjectPamphletContentDataSource.swift]: 🙌 it makes me very happy you noticed this!

File: Library/ViewModels/LiveStreamContainerViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+  var inputs: LiveStreamContainerViewModelInputs { get }
+  var outputs: LiveStreamContainerViewModelOutputs { get }
+}
+
+public protocol LiveStreamContainerViewModelInputs {
+  func configureWith(project: Project, event: LiveStreamEvent?)
+  func closeButtonTapped()
+  func liveStreamViewControllerStateChanged(state: LiveStreamViewControllerState)
+  func retrievedLiveStreamEvent(event: LiveStreamEvent)
+  func viewDidLoad()
```
Comment (theginbin) [inline Library/ViewModels/LiveStreamContainerViewModel.swift]: we like to add doc comments for all inputs and outputs. if you have time to add them for all view models, that'd be great, but a future PR would be fine too :)

File: Library/ViewModels/LiveStreamEventDetailsViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+      self.configData.signal.skipNil(),
+      self.viewDidLoadProperty.signal
+      )
+      .map(first)
+
+    let project = configData.map(first)
+
+    let eventEvent = configData
+      .switchMap { project, liveStream, optionalEvent in
+        fetchEvent(forProject: project, liveStream: liveStream, event: optionalEvent)
```
Comment (theginbin) [inline Library/ViewModels/LiveStreamEventDetailsViewModel.swift]: do we want to add the scheduler to this? `.ksr_delay(AppEnvironment.current.apiDelayInterval, on: AppEnvironment.current.scheduler)`

File: Library/ViewModels/LiveStreamEventDetailsViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+    self.openLoginToutViewController = self.subscribeButtonTappedProperty.signal
+      .filter { AppEnvironment.current.currentUser == nil }
+
+    self.creatorName = event.map { $0.creator.name }
+    self.liveStreamTitle = event.map { $0.stream.name }
+    self.liveStreamParagraph = event.map { $0.stream.description }
+
+    self.subscribeButtonImage = subscribed.map {
+      $0 ? "postcard-checkmark" : nil
+    }
```
Comment (theginbin) [inline Library/ViewModels/LiveStreamEventDetailsViewModel.swift]: tiny thing, but would this be better to just show/hide the image instead of setting to nil?

File: Library/ViewModels/LiveStreamEventDetailsViewModel.swift [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+      .observeValues { configData, isSubscribed in
+        let (project, liveStream, _) = configData
+        AppEnvironment.current.koala.trackLiveStreamToggleSubscription(project: project,
+                                                                       liveStream: liveStream,
+                                                                       subscribed: isSubscribed)
+    }
+  }
+
+  private let configData = MutableProperty<(Project, Project.LiveStream, LiveStreamEvent?)?>(nil)
+  public func configureWith(project: Project, liveStream: Project.LiveStream, event: LiveStreamEvent?) {
```
Comment (theginbin) [inline Library/ViewModels/LiveStreamEventDetailsViewModel.swift]: feel like this could be an enum such as `LiveStreamData` then you could use dot notation to access what you need.

File: Library/ViewModels/LiveStreamEventDetailsViewModel.swift (hunk 15 above)
Comment (theginbin) [inline Library/ViewModels/LiveStreamEventDetailsViewModel.swift]: oh my b, i see you have it in a private function. ignore me!

File: LiveStream/Extensions/LiveStreamTypes.swift [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+  case (.greenRoom, .greenRoom):
+    return true
+  case (.live(let lhsPlaybackState, let lhsStartTime), .live(let rhsPlaybackState, let rhsStartTime)):
+    return lhsPlaybackState == rhsPlaybackState && lhsStartTime == rhsStartTime
+  case (.replay(let lhsPlaybackState, let lhsDuration), .replay(let rhsPlaybackState, let rhsDuration)):
+
+    return lhsPlaybackState == rhsPlaybackState && lhsDuration == rhsDuration
+
+  case (.error(let lhsError), .error(let rhsError)):
+
```
Comment (theginbin) [inline LiveStream/Extensions/LiveStreamTypes.swift]: check the extra spaces from lines 101-105

File: LiveStream/LiveStream.h [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+#import <UIKit/UIKit.h>
+
+//! Project version number for LiveStream.
+FOUNDATION_EXPORT double LiveStreamVersionNumber;
+
+//! Project version string for LiveStream.
+FOUNDATION_EXPORT const unsigned char LiveStreamVersionString[];
+
+// In this header, you should import all the public headers of your framework using statements like #import <LiveStream/PublicHeader.h>
+
```
Comment (theginbin) [inline LiveStream/LiveStream.h]: is this file a mistake? objc? if not, remove Brando credit at the top and instructions at the bottom.

File: LiveStream/Models/LiveStreamEvent.swift [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+    public let name: String
+    public let projectWebUrl: String
+    public let projectName: String
+    public let replayUrl: String?
+    public let startDate: Date
+    public let webUrl: String
+
+    // Useful for safeguarding against getting a `hasReplay == true` yet the `replayUrl` is `nil`.
+    public var definitelyHasReplay: Bool {
+      return self.hasReplay && self.replayUrl != nil
```
Comment (theginbin) [inline LiveStream/Models/LiveStreamEvent.swift]: is it possible to **have** a `replayUrl`, but **not** have replay enabled (`hasReplay = false`)? asking because then you could just do the check on the `replayUrl` and ignore `hasReplay`?

File: LiveStream/Models/LiveStreamEvent.swift [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+private let dateFormatter: DateFormatter = {
+  let dateFormatter = DateFormatter()
+  dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZ"
+  return dateFormatter
+}()
+
+private func toDate(dateString: String) -> Decoded<Date> {
+
+  guard let date = dateFormatter.date(from: dateString) else {
+    return .failure(DecodeError.custom("Unable to parse date format"))
```
Comment (theginbin) [inline LiveStream/Models/LiveStreamEvent.swift]: so i can't remember - does this break the decoding if it's a failure? do we want this?

File: LiveStream/Service/KsLive.swift [hunk 22]
```diff
@@ -0,0 +1,39 @@
+import FirebaseAnalytics
+import FirebaseDatabase
+import FirebaseAuth
+
+public class KsLiveApp {
```
Comment (theginbin) [inline LiveStream/Service/KsLive.swift]: do you want to rename this file to match the class name? `KsLive` -> `KsLiveApp`

File: LiveStream/Models/LiveStreamEvent.swift (hunk 21 above)
Comment (mbrandonw) [inline LiveStream/Models/LiveStreamEvent.swift]: it does indeed :( ideally we'd send back dates as doubles, but the live stream API sends it back in the iso date string format. there isn't really anything we can do other than change the live stream api, which maybe we can do sometime soon.

File: LiveStream/Models/LiveStreamEvent.swift (hunk 20 above)
Comment (mbrandonw) [inline LiveStream/Models/LiveStreamEvent.swift]: 🤔 we're not sure! we were seeing some funky data coming back from the live stream api, and so we are just trying to be super explicit.

ostensibly there is a chance that maybe the `replayUrl` exists cause it's basically an archive of the stream, but perhaps the creator didnt want a replay to be available? or there were technical difficulties?

/cc @Jwomers

File: LiveStream/LiveStream.h (hunk 19 above)
Comment (mbrandonw) [inline LiveStream/LiveStream.h]: ha good catch! btw, everything is copywrited under brando inc.

File: Library/ViewModels/LiveStreamContainerViewModel.swift (hunk 14 above)
Comment (justinswart) [inline Library/ViewModels/LiveStreamContainerViewModel.swift]: Ah let me do that!

File: LiveStream/Service/KsLive.swift (hunk 22 above)
Comment (mbrandonw) [inline LiveStream/Service/KsLive.swift]: actually ended up deleting and moving stuff to the `LiveStreamServiceProtocol` which cleans up things!

File: Library/ViewModels/LiveStreamEventDetailsViewModel.swift (hunk 17 above)
Comment (justinswart) [inline Library/ViewModels/LiveStreamEventDetailsViewModel.swift]: Yeah this makes sense, I'm going to add this to our Trello and do as a separate PR 👍

File: LiveStream/Service/KsLive.swift (hunk 22 above)
Comment (justinswart) [inline LiveStream/Service/KsLive.swift]: Oh so that's an interesting observation, I initially tried to do that when `KsLive` was a separate framework, however this caused ambiguity when I wanted to call `KsLive.something()` from the main app so I had to rename the class to something that was not the same name as the framework. @mbrandonw has since refactored things so that this class doesn't exist at all but just thought I'd mention why it was there in the first place 😄

File: Library/ViewModels/LiveStreamEventDetailsViewModel.swift (hunk 16 above)
Comment (justinswart) [inline Library/ViewModels/LiveStreamEventDetailsViewModel.swift]: Unfortunately this doesn't work with the way that I'm using the subscribe button's image - I'm placing a checkmark to the right of the text using `self.subscribeButton.semanticContentAttribute = .forceRightToLeft`. Setting the button's `imageView.isHidden` to `true` doesn't seem to work and I also think if it did work it may not clear the space properly.
